---
title: "Step 4: Results"
subtitle: "Hunting DNA methylation repression of transcriptional activity"
author: "Florence Pittion, Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---

# Step 4 : Application to cancers

```{r, echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide", warning=FALSE)
source("params.R")
```

```{r}
# faire boucle pour les deux (voir 3) TCGA 

if (!exists("momik_pvalbetas")) {momik_pvalbetas = list()}

s_meth = mreadRDS("~/projects/all_human_tissues/results/meth/study_meth_all_human_tissues_grch38.rds")
for (tcga_project in tcga_projects) {
  print(tcga_project)
  if (!tcga_project %in% names(momik_pvalbetas)) {
    genes_singleton = mget_genes(tcga_project)
    genes = genes_singleton$genes
    feat_indexed_probes = mget_feat_indexed_probes(feats_bed6=genes, s_meth$platform[,c("Chromosome", "Start")], up_str=ud_str, dwn_str=2500)
    tmp_gs = intersect(names(feat_indexed_probes), rownames(genes))
    
    momik_pvalbetas_tmp = epimedtools::monitored_apply(mod=10, t(t(tmp_gs)), 1, function(gene_symbol) {
      #print(gene_symbol)
      feat = genes[gene_symbol,]
      tss = ifelse(feat[[6]]=="+", feat[[2]], feat[[3]])
      interaction_range=2500
      region_id = paste0(feat[[1]], ":", tss-interaction_range, "-", tss+interaction_range)      
      feat_indexed_probes[[region_id]] = feat_indexed_probes[[gene_symbol]]
      mdata = try(mget_multiomic_data(region_id=region_id, gene_symbols=gene_symbol, tcga_project=tcga_project, feat_indexed_probes=feat_indexed_probes))   
      if (class(mdata) != "try-error"){
        if ((!is.null(mdata$d)) & (gene_symbol %in% colnames(mdata$d))) {
          mdata$d 
          head(mdata$d)
          Y = mdata$d[,gene_symbol]
          X = apply(mdata$d[,mdata$probes],1,mean, na.rm=TRUE)
          m = lm(Y~X)
          
          sm =summary(m)
          fstat=sm$fstatistic
          pval = pf(fstat[[1]], fstat[[2]], fstat[[3]], lower.tail = FALSE)
          beta = m$coefficient[[2]]
          return(c(beta=beta, pval=pval))
        } else {
          warning(paste0("probleme with ", gene_symbol))
          return(NULL)
        }
      } else {
        warning(paste0("probleme with ", gene_symbol))
        return(NULL)
      }
    })  
    names(momik_pvalbetas_tmp) = tmp_gs  
    momik_pvalbetas_tmp = do.call(rbind,momik_pvalbetas_tmp)
    momik_pvalbetas[[tcga_project]] = momik_pvalbetas_tmp
  }
  tmp_idx = intersect(rownames(feats),rownames(momik_pvalbetas[[tcga_project]]))
  feats[,paste0("momik_betas_", tcga_project)] = NA
  feats[tmp_idx,paste0("momik_betas_", tcga_project)] = momik_pvalbetas[[tcga_project]][tmp_idx,1]
  feats[,paste0("momik_pvals_", tcga_project)] = NA
  feats[tmp_idx,paste0("momik_pvals_", tcga_project)] = momik_pvalbetas[[tcga_project]][tmp_idx,2]
  feats[,paste0("momik_lpvals_", tcga_project)] = -log10(feats[,paste0("momik_pvals_", tcga_project)])
}
#feats$momik_pvals = NA
#feats[rownames(momik_pvalbetas),]$momik_pvals = momik_pvalbetas[,2]
#feats$momik_betas = NA
#feats[rownames(momik_pvalbetas),]$momik_betas = momik_pvalbetas[,1]
#feats$momik_lpvals = -log10(feats$momik_pvals)
```

## Figure


```{r fig.height=6}
for (tcga_project in tcga_projects) {
  layout_mat2 =
    cbind(
      matrix(2, nrow=3, ncol=1),
      matrix(1, nrow=3, ncol=4),
      matrix(3, nrow=3, ncol=3)
    )

  layout(layout_mat2, respect=TRUE)

  foo = momic_pattern("TSPYL5", tcga_project, LAYOUT=FALSE)
  fig_label("A", cex=3)

  meth = apply(foo$d[,foo$probes], 1, mean)
  expr = foo$d[,foo$gene_symbols]
  plot(meth, expr, main=paste0("TSPYL5 expr. ~ av. meth."), xlab="average methylation", ylab= "expression")
  m = lm(expr~meth)
  abline(m, col=2)
  fig_label("B", cex=3)

  legend("bottomleft", c(paste0("beta=", signif(m$coef[[2]],3)), paste0("Fisher pval=", signif(anova(m)[1,5],3))), bty = "n")
}
```




```{r fig.height=9, eval=TRUE}
ntop=100
idx = rownames(feats)[feats[["methplusplus"]]]
idx2 = idx[order(feats[idx, ]$l2fc_GSE45332, decreasing = TRUE)][1:ntop]
idx3 = idx[order(feats[idx, ]$l2fc_GSE5816, decreasing = TRUE)][1:ntop]
gene_symbols = (union(idx2, idx3))
gene_symbols = gene_symbols[order(feats[gene_symbols, paste0("momik_pvals_", tcga_project)])][1:min(6, length(gene_symbols))]

for (tcga_project in tcga_projects) {
  layout(matrix(1), respect=TRUE)
  feats[,paste0("momik_lpvals_", tcga_project)] = -log10(feats[,paste0("momik_pvals_", tcga_project)])
  plot(feats[,paste0("momik_betas_", tcga_project)], feats[,paste0("momik_lpvals_", tcga_project)], col="grey", 
    xlim=c(-40,40),
    main=paste0(tcga_project, " gene expression ~ TSS+/-2.5kb methylation"), xlab="beta (linear regression coefficient)", ylab="-log10(Fisher p-value)"
  )
  points(feats[idx, paste0("momik_betas_", tcga_project)], feats[idx, paste0("momik_lpvals_", tcga_project)], col="red")
  points(feats[idx2,paste0("momik_betas_", tcga_project)], feats[idx2,paste0("momik_lpvals_", tcga_project)], col="blue", pch=16)
  points(feats[idx3,paste0("momik_betas_", tcga_project)], feats[idx3,paste0("momik_lpvals_", tcga_project)], col="purple", pch=16)
  text(feats[gene_symbols,paste0("momik_betas_", tcga_project)], feats[gene_symbols,paste0("momik_lpvals_", tcga_project)], gene_symbols, col=adjustcolor(1, alpha.f=1), pos=3)
  legend("topright", 
    c("genes", paste0("methplusplus", " genes"), paste0("methplusplus", " genes upregulated GSE45332 DA"), paste0(gs_name, " genes upregulated GSE5816 DA")), 
    pch=c(1, 1, 16, 16), 
    col=c("grey", "red", "blue", "purple"),
    lty=c(0,0,0,0)
  )
}

```



```{r}
for (tcga_project in tcga_projects) {
  layout(matrix(1:6 , 2), respect=TRUE)
  gene_set = rownames(feats)[feats[["methplusplus"]]]
  expression_vector = feats[,paste0("momik_betas_", tcga_project)]
  names(expression_vector) = rownames(feats)
  expression_vector = expression_vector[!is.na(expression_vector)]
  et_gsea_plot(expression_vector, gene_set, prefix="methplusplus", nperm=3)  

  for (gse in gses) {
    idx = rownames(feats)[feats[["methplusplus"]]]
    idx2 = idx[order(feats[idx,paste0("l2fc_", gse)], decreasing=TRUE)][1:ntop]
    gene_set = idx2
    expression_vector = feats[idx,paste0("momik_betas_", tcga_project)]
    names(expression_vector) = idx
    expression_vector = expression_vector[!is.na(expression_vector)]
    et_gsea_plot(expression_vector, gene_set, prefix=paste0("top", ntop, "_l2fc_", gse), nperm=3)
  }
}
```

**Correlation between gene expression and TSS+/-2.5kb methylation in the TCGA-LUSC dataset (Lung Squamous Cell Carcinoma samples).** 
For each genes a linear regression of the expression as a function of the average methylation of its TSS+/-2.5kb was performed. 
**(A)** As an example,  a heatmap shows the methylation levels (scale ranging between blue corresponding to lowest and red to highest methylation levels) of the of CpGs of the TSPYL5 gene promoter (x-axis) for the TCGA-LUSC samples (y-axis) ranked according to the expression of the same gene (ranked from high (top) to low (bottom) expression , y-axis). This TSPYL5 TSS+/-2.5kb methylation heatmap shows that a low TSPYL5 expression is associated to a strong methylation (in red) of the region surrounding TSPYL5 TSS whereas high expression of TSPYL5 is associated to a demethylation of the same genomic region (in blue). 
**(B)** The average methylation value of TSPYL5 promoter region was calculated for each patient and linear regression *expression ~ average methylation* was performed. The TSPYL5 expression (y-axis) is plotted as a function of the TSPYL5 average methylation (x=axis) and each point represent a sample. The red line is the regression line. The coefficient of the linear model (beta=$`r signif(m$coef[[2]],3)`$) and the associated Fisher p-value $`r signif(anova(m)[1,5],3)`$ are reported on the graph.
**(C)** In the same way, beta and p-values were collected for all genes and plotted on the corresponding volcano plot (one dot/gene, the x-axis represents the linear regression coefficient of the model and y-axis represents -log10 of the Fisher p-value). Genes are represented as grey circles except *methplusplus* genes represented as red circles. The top `r ntop` upregulated genes of the model `r models[[1]]` (resp. `r models[[2]]`) in gse `r gses[[1]]` (resp. `r gses[[2]]`) dataset are represented as blue (resp. purple) dots. The plot shows that for most of the genes an increased expression correlates with a decrease of their TSS+/-2.5kb methylation.
**(D)** The GSEA enrichment plots show the **depletion** of ***methplusplus* genes** in the XXX-length **all genes** vector of linear regression coefficients. It means that the epression of *methplusplus* genes anticorrelated with the methylation status of the region surrounding their TSS.
**(E)** (resp. **(F)**) The enrichment plot shows the depletion of the  **top `r ntop` upregulated genes** of the model `r models[[1]]` (resp. `r models[[2]]`) in the XXX-length  ***methplusplus* genes** vector of linear regression coefficients.
It means that the expression of *methplusplus* genes that are reactivated in a demethylating context is more significantly associated with a demethylation of the region surrounding their TSS than *methplusplus* genes themselves.




## Discussion

Nicotine promotes the development of non-small cell lung cancer through activating LINC00460 and PI3K/Akt signaling.
Zhao H, Wang Y, Ren X.
Biosci Rep. 2019 Jun 7;39(6):BSR20182443. doi: 10.1042/BSR20182443. Print 2019 Jun 28.
PMID: 31123168 Free PMC article.

LINC00460 promotes proliferation and inhibits apoptosis of non-small cell lung cancer cells through targeted regulation of miR-539.
Wang HX, Kang LJ, Qin X, Xu J, Fei JW.
Eur Rev Med Pharmacol Sci. 2020 Jun;24(12):6752-6758. doi: 10.26355/eurrev_202006_21663.
PMID: 32633366 Free article.

Long non-coding RNA linc00460 promotes epithelial-mesenchymal transition and cell migration in lung cancer cells.
Li K, Sun D, Gou Q, Ke X, Gong Y, Zuo Y, Zhou JK, Guo C, Xia Z, Liu L, Li Q, Dai L, Peng Y.
Cancer Lett. 2018 Apr 28;420:80-90. doi: 10.1016/j.canlet.2018.01.060. Epub 2018 Jan 31.
PMID: 29409808


Linc00221 modulates cisplatin resistance in non-small-cell lung cancer via sponging miR-519a.
Tang H, et al. Biochimie, 2019 Jul. PMID 31029744

Mitochondria localization and dimerization are required for CIDE-B to induce apoptosis.
Chen Z, Guo K, Toh SY, Zhou Z, Li P.
J Biol Chem. 2000 Jul 28;275(30):22619-22. doi: 10.1074/jbc.C000207200.
PMID: 10837461

Dual promoters control the cell-specific expression of the human cell death-inducing DFF45-like effector B gene.
Da L, Li D, Yokoyama KK, Li T, Zhao M.
Biochem J. 2006 Feb 1;393(Pt 3):779-88. doi: 10.1042/BJ20051027.
PMID: 16248853 Free PMC article.

Targeting therapy-resistant lung cancer stem cells via disruption of the AKT/TSPYL5/PTEN positive-feedback loop
In-Gyu Kim, Jei-Ha Lee, Seo-Yeon Kim, Chang-Kyu Heo, Rae-Kwon Kim & Eun-Wie Cho 
Communications Biology volume 4, Article number: 778 (2021) Cite this article

TSPYL5 is involved in cell growth and the resistance to radiation in A549 cells via the regulation of p21(WAF1/Cip1) and PTEN/AKT pathway
Eun Jin Kim  1 , So Yong Lee, Tae Rim Kim, Soo Im Choi, Eun Wie Cho, Kug Chan Kim, In Gyu Kim




## Export features


```{r}
feat_export_filename = paste0("feats_", prefix3, ".xlsx")
```

Annotated features are exported as [`r feat_export_filename`](`r feat_export_filename`) file. 

```{r results="verbatim"}
head(feats)
```

```{r export_genes}
WriteXLS::WriteXLS(feats, feat_export_filename)
```






















```{r survival, eval=FALSE}
## Survival

if (! exists("survival_pvalhrs")) {
  s = mreadRDS(paste0("~/projects/tcga_studies/study_", tcga_project, "_trscr.rds"))
  s = truncate_survival(s, 60)  

  genes_singleton = mget_genes(tcga_project)
  genes = genes_singleton$genes

  tmp_gs = intersect(rownames(feats), rownames(s$data))
  survival_pvalhrs = epimedtools::monitored_apply(mod=10, t(t(tmp_gs)), 1, function(gene_symbol) {
    # gene_symbol = "MIR6723"
    # print(gene_symbol)
    idx_sample = rownames(s$exp_grp)[!is.na(s$exp_grp$os)]
    ss = s$exp_grp[idx_sample,]$os
    v = s$data[gene_symbol, idx_sample]
    pv = try(epimedtools::coxres(ss=ss, v=v)[[1]])
    if (class(pv)!="try-error") {
      hr = epimedtools::coxres(ss=ss, v=v)[[4]]
    } else {
      pv = 1
      hr = 1
    }
    ret = c(hr=hr, pv=pv)
    return(ret)
  })  
  survival_pvalhrs = t(survival_pvalhrs)
  rownames(survival_pvalhrs) = tmp_gs
  survival_pvalhrs[is.na(survival_pvalhrs[,1]),1] = 1
}
feats$survival_pvals = NA
feats[rownames(survival_pvalhrs),]$survival_pvals = survival_pvalhrs[,2]
feats$survival_hrs = NA
feats[rownames(survival_pvalhrs),]$survival_hrs = survival_pvalhrs[,1]
feats$survival_lpvals = -log10(feats$survival_pvals)

# feats[,paste0("survival_pvals_", tcga_project)] = NA
# feats[,paste0("survival_pvals_", tcga_project)] = survival_pvalhrs[rownames(feats),2]
# feats[,paste0("survival_hrs_", tcga_project)] = NA
# feats[,paste0("survival_hrs_", tcga_project)] = survival_pvalhrs[rownames(feats),1]













gene_symbols = intersect(intersect(idx[feats[idx,]$survival_pvals < 0.05], idx[-log10(feats[idx,]$momik_pvals)>5]), union(idx2, idx3)) 
gene_symbols = gene_symbols[order(feats[gene_symbols,]$momik_pvals)][1:min(20, length(gene_symbols))]


layout(1, respect=TRUE)
plot(-log10(feats$momik_pvals), -log10(feats$survival_pvals), col=adjustcolor(1, alpha.f=.3), 
  xlab="multiomic score", 
  ylab="-log10(survival logrank p-value)", xlim=c(0,2*max(-log10(feats[gene_symbols,]$momik_pvals))), ylim=c(0,2*max(-log10(feats[gene_symbols,]$survival_pvals))))
abline(h=-log10(0.05), lty=2)
points(-log10(feats[idx,]$momik_pvals), -log10(feats[idx,]$survival_pvals), col="red")
points(-log10(feats[idx2,]$momik_pvals), -log10(feats[idx2,]$survival_pvals), col=adjustcolor("blue", alpha.f=.8), pch=16)
points(-log10(feats[idx3,]$momik_pvals), -log10(feats[idx3,]$survival_pvals), col=adjustcolor("purple", alpha.f=.8), pch=16)

# text(-log10(feats[gene_symbols,]$momik_pvals), -log10(feats[gene_symbols,]$survival_pvals), gene_symbols, pos=3, col="white")
text(-log10(feats[gene_symbols,]$momik_pvals), -log10(feats[gene_symbols,]$survival_pvals), gene_symbols, col=adjustcolor(1, alpha.f=1), pos=c(1,3))
legend("topright", 
  c("genes", paste0(gs_name, " genes"), paste0(gs_name, " genes upregulated GSE45332 DA"), paste0(gs_name, " genes upregulated GSE5816 DA"), "5% survival logrank p-value threshold "), 
  pch=c(1, 1, 16, 16, NA), 
  col=c("grey", "red", "blue", "purple", "black"),
  lty=c(0,0,0,0,2)
)


tcga_project = "TCGA-LUSC"
s = mreadRDS(paste0("~/projects/tcga_studies/study_", tcga_project, "_trscr.rds"))
s = truncate_survival(s, 60)
idx_sample = rownames(s$exp_grp)[!is.na(s$exp_grp$os)]


foo = sapply(gene_symbols, function(gene_symbol) {
  ss = s$exp_grp[idx_sample,]$os
  v = s$data[gene_symbol, idx_sample]
  par(mar=c(2.5, 2.5, 4.1, 2.1))
  ret = try(epimedtools::plot_survival_panel_simple2(ss=ss, v=v, main=gene_symbol))
  try(momic_pattern(gene_symbol, tcga_project))
  ret
})

# thresh = do.call(rbind,foo[1,])[,2]
#
# idx_genes = gene_symbols[1:4]
# v = apply(s$data[idx_genes,idx_sample] > thresh[idx_genes], 2, sum)
# table(v)
# ss = s$exp_grp[idx_sample,]$os
# par(mar=c(2.5, 2.5, 4.1, 2.1))
# ret = try(epimedtools::plot_survival_panel_simple2(ss=ss, v=v, main=gene_symbol))

```














```{r core feats, eval=FALSE}
## Core feats
# plot 2 firsts studies
#threshold
#study 1
i = 1
#study 2
j = 1
layout(1, respect=TRUE)
plot(feats$l2fc_GSE45332, feats$l2fc_GSE5816, xlab="l2fc GSE45332 WTvsDKO", ylab="l2fc GSE5816", pch=".", col="grey")
idx = rownames(feats)[!is.na(feats$methplusplus) & feats$methplusplus]
points(feats[idx,]$l2fc_GSE45332, feats[idx,]$l2fc_GSE5816, col=2, pch=1)
abline(h=i, col=1, lty=2)
abline(v=j, col=1, lty=2)
legend("topleft", pch=1, col=2, "gs_methplusplus")

# features of interest
corefeats = rownames(feats)[
  !is.na(feats$l2fc_GSE45332) & feats$l2fc_GSE45332>i & 
  !is.na(feats$l2fc_GSE5816)  & feats$l2fc_GSE5816>j  & 
  !is.na(feats$methplusplus)  & feats$methplusplus
]
length(corefeats) 
```


```{r results="verbatim", eval=FALSE}
# Annotated core features are exported as [`r corefeat_export_filename`](`r corefeat_export_filename`) file. 
if (length(corefeats!=0)){
  corefeat_export_filename = paste0("corefeats_", prefix3, ".xlsx")
  head(corefeats)
  tmp_d = feats[corefeats,]
  WriteXLS::WriteXLS(tmp_d, corefeat_export_filename)
}
```




```{r gtex, fig.height=9, eval=FALSE}
## Expression in GTEx
expression_in_gtex(gene_symbols)
```






