---
title: "Step 1: CpG status of region surrounding TSS"
author: "Florence Pittion, Florent Chuffart"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
  pdf_document:
    keep_tex: true
---



```{r, echo=FALSE, eval=TRUE, label="loading libraries"}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
```

```{r}
source("common.R")
if (!exists("stp")) stp = 100
if (!exists("nb_rnd_feat")) nb_rnd_feat = 2000
if (!exists("ud_str")) ud_str = 2500
up_str = dwn_str = ud_str
prefix = "tss"
prefix = paste0(prefix, "_", up_str, "_", dwn_str, "_", stp, "_", nb_rnd_feat, "")
```

```{r globals variables}
hg38_chrom_sizes  = read.table("~/projects/genes/hg38.chrom.sizes", stringsAsFactors=FALSE)
rownames(hg38_chrom_sizes) = hg38_chrom_sizes[,1]
```

# Step 1: CpG status of region surrounding TSS
## Purpose

DNA methylation is one of several epigenetic mechanisms that cells use to control gene expression. Since many cancers have low levels of DNA methylation compared to most mammalian somatic tissues, we want to investigate the implication of DNA methylation as a key regulator of gene expression contributing to cancer development . The first objective of this study is to group genes according to the three following characteristics: 

- CpG density of their promoter region, 
- level of methylation of the promoter region in normal tissues, 
- level of expression in a demethylating context. 

The aim is to define a set of genes whose promoter region is CpG-rich and widely methylated in most tissues and whose expression is up-regulated in a demethylating context. The second objective of this study is to investigate whether this subset of genes could be enriched in cancer aggressiveness biomarkers. To investigate the first question, we are developing dmethr a three-stp dedicated pipeline, and apply it on publicly available datasets.

**Data**

This study is based on following data set:

- methylation data from normal tissues (Illumina 450k: GSE56515, GSE48472, GSE64096, GSE50192", GSE31848", GSE73375)
- transcriptomic data HCT116 DNMT DKO vs. WT (RNA-seq: GSE45332) 
- transcriptomic microarray from control and 5-azacytidine treated lung cancer cell lines (GSE5816)

## Features

We use definition of transcripts proposed by UCSC, version hg38 (`genome`).



```{r}
feats_orig = readRDS("~/projects/genes/bed_grch38_epimeddb.rds")[,1:6]
feats = feats_orig
head(feats)
dim(feats)
```

```{r filtering anbd downsampling}
# Filtering according to...
# feats = feats[
#   # feats$score>5 &
#   feats$len<5000
#   ,]
# plot_feat_den(feats$len, feats$score, xlim, ylim)

# Downsampling
if (nb_rnd_feat > 0) {
  dim(feats)
  set.seed(1)
  feats = feats[sort(sample(1:nrow(feats), min(nrow(feats), nb_rnd_feat))),]
  dim(feats)  
  # plot_feat_den(feats$len, feats$score, xlim, ylim, xlab="length", ylab="#probes")
}
```


## CpG enrichment of features


```{r Extract DNA sequences of TSS +/- ud_str bp}
seq_filename = paste0("seq_", prefix, ".rds")
if (!file.exists(seq_filename)) {
  seq = get_seq_from_bed(feats, up_str, dwn_str, hg38_chrom_sizes)
  saveRDS(seq, seq_filename)
}
seq = readRDS(seq_filename)
length(seq)
```

```{r Slice DNA sequences and count CpG density}
mat = mget_matbin_from_seq(seq, up_str, dwn_str, stp)
head(mat)
dim(mat)

binmax = apply(mat, 1, function(l) {
  maxIndex = which(l%in%max(l, na.rm=TRUE))[1]
})
head(binmax)
feats$binmax = NA
feats[names(binmax),]$binmax = binmax


cpg_density = apply(mat, 1, mean, na.rm=TRUE)
feats$cpg_density = NA
feats[names(cpg_density),]$cpg_density = cpg_density


head(feats)
```


CpG density ( nb. of CpG / length of sequence) is computed in sequence of TSS +/-`r ud_str/1000`kb sliced into `r stp` bins.

TSS of 1000 random feats is represented there. It shows a CpG enrichment aoroud TSS.

```{r label="heatmap for 1000 rnd feats", fig.height=9}
set.seed(1)
data = mat[order(binmax[rownames(mat)]),]
data = data[sort(sample(1:nrow(data), min(nrow(data), 1000))), ]
colors=c("black",  "red")
cols = colorRampPalette(colors)(20)
rownames(data) = NULL
gplots::heatmap.2(data, Rowv=NULL,  Colv=NULL,  dendrogram="none", trace="none", col=cols, main=paste0("CpG enrichment (feats +/-", ud_str/1000, "kb, 1000 rnd feats)"), xaxt="n", mar=c(8,5))
```

`r if(feature_pretreatment == "cen") "## Centered features"`


```{r centered, fig.height=9}
if (feature_pretreatment == "cen") {
  # stop("EFN")
  feats$start_orig = feats[,2] 
  feats$end_orig = feats[,3]  
  feats$strand_orig = feats[,6]  
  feats$tss = ifelse(feats[,6]=="+", feats[,2], feats[,3])
  # feats[,2] = feats$tss - (feats[,6]=="-")+0
  feats[,2] = feats$tss - ud_str + stp*feats$binmax + floor(stp/2)
  feats[,3] = feats[,2] + 1
  head(feats)

  seq_filename = paste0("seq_centered_", prefix, ".rds")
  if (!file.exists(seq_filename)) {
    seq = get_seq_from_bed(feats, up_str, dwn_str, hg38_chrom_sizes)
    saveRDS(seq, seq_filename)
  }
  seq = readRDS(seq_filename)
  length(seq)
  # ```
  # ```{r mat_cpg_binmax}
  mat = mget_matbin_from_seq(seq, up_str, dwn_str, stp)
  head(mat)
  dim(mat)
  binmax = apply(mat, 1, function(l) {
    maxIndex = which(l==max(l))[1]
  })
  head(binmax)
  feats$binmax = NA
  feats[names(binmax),]$binmax = binmax
  head(feats)
  # ```
  # ```{r label="heatmap for 1000 rnd feats", fig.height=9}
  set.seed(1)
  data = mat[order(binmax[rownames(mat)]),]
  data = data[sort(sample(1:nrow(data), min(nrow(data), 1000))), ]
  colors=c("black",  "red")
  cols = colorRampPalette(colors)(20)
  rownames(data) = NULL
  gplots::heatmap.2(data, Rowv=NULL,  Colv=NULL,  dendrogram="none", trace="none", col=cols, main=paste0("CpG enrichment (feats +/-", ud_str/1000, "kb, 1000 rnd feats)"), xaxt="n", mar=c(8,5))  
} 
```


## CpG density


CpG density distribution is represented there. 

It shows that MOST OF TSS are CpG enriched compared to random sequence.  

It also shows that NOT ALL TSS are are CpG enriched.

```{r label="CpG density distribution"}
nb_cpg_rnd_filename = paste0("nb_cpg_rnd_TSS_", up_str, "_", dwn_str, ".rds")
if (!file.exists(nb_cpg_rnd_filename)) {
  idx = sample(1:22, 5000, replace=TRUE)
  set.seed(1)
  pos = sapply(hg38_chrom_sizes[idx,2], function(e) round(runif(1, 15000, e-15000)))
  rnd_feat = data.frame(hg38_chrom_sizes[idx,1], pos, stringsAsFactors=FALSE)
  nb_cpg_rnd = epimedtools::monitored_apply(mod=10, rnd_feat, 1, function(feat) {
    chr = feat[[1]]
    tss = as.numeric(feat[[2]])
    bef = up_str
    aft = dwn_str
    beg = max(1,tss - bef)
    end = min(hg38_chrom_sizes[chr,2], tss + aft)
    str = as.character(BSgenome::getSeq(BSgenome.Hsapiens.UCSC.hg38::Hsapiens, chr, beg, end))
    # stringr::str_count(str, "CG") / nchar(str)
    lnn = nchar(gsub("N", "", str))
    # denom = l - lnn
    if (lnn==0) {
      return(NA)
    } else {
      return(stringr::str_count(str, "CG") / lnn)
    }
  })
  saveRDS(nb_cpg_rnd, nb_cpg_rnd_filename)
}
nb_cpg_rnd = readRDS(nb_cpg_rnd_filename)

nb_cpg_tss = apply(mat, 1, mean, na.rm=TRUE)

layout(matrix(1:2,1), respect=TRUE)
plot(density(nb_cpg_tss),ylim=c(0,100), main="CpG enrichment distribution", xlab="CpG density", ylab="", yaxt="n")
lines(density(nb_cpg_rnd, na.rm=TRUE), col=2)
legend("topright", c(paste0("tss +/-", ud_str/1000, "kb"), paste0("rnd pos +/-", ud_str/1000, "kb")), col=1:2, lty=1)
plot(density(log10(nb_cpg_tss)), main="CpG enrichment distribution", xlab="log10(CpG density)", ylab="", yaxt="n")
lines(density(log10(nb_cpg_rnd), na.rm=TRUE), col=2)
legend("topright", c(paste0("tss +/-", ud_str/1000, "kb"), paste0("rnd pos +/-", ud_str/1000, "kb")), col=1:2, lty=1)
```



## Clustering 


Then we clustering TSS according to there sliced CpG density

PCA on TSS +/-`r ud_str/1000`kb sliced into `r stp`b bins is performed.

Component 1 discriminates CpG rich and poor TSS.

```{r clustering}
mat = mat[!apply(is.na(mat), 1, any),] 
pca = prcomp(mat, scale=FALSE)
v = pca$sdev * pca$sdev
p = v / sum(v) * 100


layout(matrix(1:2,1), respect=TRUE)
barplot(p)
i = 1
plot(density(pca$x[,i]), main=paste0("PC", i, "(", signif(p[i], 3), "%)"))

#set.seed(3)
#km = kmeans(pca$x[,1], 2)
#thresh = mean(km$centers)
#abline(v=thresh)

#plot(pca$x[,1], pca$x[,2], col=km$cluster)
#abline(v=thresh)

#idx1 = names(km$cluster)[km$cluster==1]
#idx2 = names(km$cluster)[km$cluster==2]

library(mclust)
em = Mclust(pca$x[,1], G=2)
plot(em, what = "density")
summary(em)


idx1 = names(em$classification)[em$classification==1]
idx2 = names(em$classification)[em$classification==2]


feats$cpg_status  = NA
if (mean(mat[idx1,]) < mean(mat[idx2,])) {
  feats[idx1,]$cpg_status = "poor"
  feats[idx2,]$cpg_status = "rich"  
} else {
  feats[idx1,]$cpg_status = "rich"
  feats[idx2,]$cpg_status = "poor"
}



idx_rich = rownames(feats)[feats$cpg_status%in%"rich"]
idx_poor = rownames(feats)[feats$cpg_status%in%"poor"]

plot(pca$x[,1], pca$x[,2], col=feats[rownames(pca$x),]$cpg_status%in%"rich"+1)

plot(apply(mat[idx_poor,], 2, mean), type="l", ylim=c(0, .13), col=1, main="CpG enrichment", xaxt="n")
lines(apply(mat[idx_rich,], 2, mean), col=2)
axis(1, at=c(0, ncol(mat)/2, ncol(mat)), label=c(paste0("TSS -", ud_str/1000, "kb"), "TSS", paste0("TSS +", ud_str/1000, "kb")))
legend("topright", legend=c(paste0("rich (", length(idx_rich), ")"), paste0("poor (", length(idx_poor), ")")), col=2:1, lty=1)

feats[rownames(mat),]$cpg_density = apply(mat, 1, mean)
# WriteXLS::WriteXLS(feats, "feats_cpg_status.xls")
```








```{r}
i=1
for (j in 2:3) {
  threshs_i = quantile(pca$x[,i], c(1/3, 2/3))
  threshs_j = quantile(pca$x[,j], c(1/3, 2/3))

  layout(1, respect=TRUE)
  plot(pca$x[,i], pca$x[,j], xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"), pch=16, col=feats[rownames(pca$x),]$cpg_status%in%"rich"+1)
  abline(v=threshs_i, h=threshs_j, col="grey", lty=2)

  layout(matrix(1:9, 3, byrow=FALSE), respect=TRUE)


  idx_i = idx_j = list()
  idx_i[[1]] = rownames(pca$x)[pca$x[,i] <  threshs_i[1]]
  idx_i[[2]] = rownames(pca$x)[pca$x[,i] >= threshs_i[1] & pca$x[,i] < threshs_i[2]]
  idx_i[[3]] = rownames(pca$x)[pca$x[,i] >= threshs_i[2] ]
  idx_j[[1]] = rownames(pca$x)[pca$x[,j] <  threshs_j[1]]
  idx_j[[2]] = rownames(pca$x)[pca$x[,j] >= threshs_j[1] & pca$x[,j] < threshs_j[2]]
  idx_j[[3]] = rownames(pca$x)[pca$x[,j] >= threshs_j[2] ]

  for (it in 1:3) {
    for (jt in 1:3) {
      idx = intersect(idx_i[[it]], idx_j[[jt]])
      plot(apply(mat[idx,], 2, mean),  col=1, type="l", ylim=c(0, .13), main=paste0("PC", i, "Q", it, " PC", j, "Q" , jt), xaxt="n", ylab="CpG density")
      axis(1, at=c(0, ncol(mat)/2, ncol(mat)), label=c(paste0("TSS -", ud_str/1000, "kb"), "TSS", paste0("TSS +", ud_str/1000, "kb")))
    }
  }  
}
```



## Figure

```{r, fig.height=9}
gs_cpgrich = rownames(feats)[!is.na(feats$cpg_status) & feats$cpg_status=="rich"]
gs = rownames(feats)

par(mar=c(5.1, 4.1, 4.1, 2.1))

layout_mat = 
cbind(
  rbind(
    matrix(1, nrow=3, ncol=3),
    rbind(
      c(3,3,2),
      c(5,5,4),
      c(5,5,4)
    )
  ),
  rbind(
    matrix(6, nrow=2, ncol=3),
    matrix(7, nrow=4, ncol=3)
  )
)
layout_mat
layout(layout_mat, respect=TRUE)

par(mar=c(5.1, 4.1, 4.1, 2.1))
plot(density(nb_cpg_tss),ylim=c(0,100), main="DNA sequence average CpG density distribution", xlab="CpG density", ylab="", yaxt="n", xlim=c(0, .08))
lines(density(nb_cpg_rnd, na.rm=TRUE), col=4)
legend("topright", c(paste0("tss +/-", ud_str/1000, "kb sequences"), paste0(2*ud_str/1000, "kb hg38 random sequences")), col=c(1,4), lty=1)
fig_label("A", cex=3)



par(mar=c(2.5, 2.5, 4.1, 2.1))
foo = barplot(p[1:10], yaxt="n")
axis(1, at=nrow(foo)/2, "% explained var. by PC", tick=FALSE)
axis(2, at=c(0, 10, 20, 30, 40), c(0, 10, 20, 30, 40), las=2)
par(mar=c(0, 4.1, 4.1, 0))
i = 1
den=density(pca$x[,i])
  plot(den, main="EM clustering", xlab="", ylab="", xaxt="n", yaxt="n", col="grey")
abline (v=max(pca$x[idx_rich,i]), lty="dotted")
den_rich = density(pca$x[idx_rich,i], na.rm=TRUE, bw=den$bw)  
den_poor = density(pca$x[idx_poor,i], na.rm=TRUE, bw=den$bw)
lines(den_rich$x, den_rich$y * (den_rich$n/nrow(pca$x)), col=2, lwd=2)
lines(den_poor$x, den_poor$y * (den_poor$n/nrow(pca$x)), col=1, lwd=2)
fig_label("B", cex=3)

par(mar=c(5.1, 0, 0, 2.1))
i = 2
den = density(pca$x[,i])
den_rich = density(pca$x[idx_rich,i], na.rm=TRUE, bw=den$bw)
den_poor = density(pca$x[idx_poor,i], na.rm=TRUE, bw=den$bw)
plot(den$y, den$x, type="l", xlab="", ylab="", xaxt="n", yaxt="n", col="grey")
lines(den_rich$y * (den_rich$n/nrow(pca$x)), den_rich$x, col=2, lwd=2)
lines(den_poor$y* (den_poor$n/nrow(pca$x)), den_poor$x , col=1, lwd=2)


par(mar=c(5.1, 4.1, 0, 0))
plot(pca$x[,1], pca$x[,2], col=adjustcolor(feats[rownames(pca$x),]$cpg_status%in%"rich"+1, alpha.f=.5), xlab=paste0("PC", 1, "(", signif(p[1], 3), "%)"), ylab=paste0("PC",2, "(", signif(p[2], 3), "%)"))

par(mar=c(5.1, 4.1, 4.1, 2.1))
plot(apply(mat[idx_poor,], 2, mean), type="l", ylim=c(0, .13), col=1, main="Cluster average CpG density profiles", xaxt="n", ylab="CpG density", xlab="")
lines(apply(mat[idx_rich,], 2, mean), col=2)
axis(1, at=c(0, ncol(mat)/2, ncol(mat)), label=c(paste0("TSS -", ud_str/1000, "kb"), "TSS", paste0("TSS +", ud_str/1000, "kb")))
legend("topright", legend=c(paste0("rich (", length(idx_rich), ")"), paste0("poor (", length(idx_poor), ")")), col=2:1, lty=1)
fig_label("C", cex=3)

par(mar=c(5.1, 4.1, 4.1, 2.1))
# plot(0, main="heatmap")





data = mat[order(feats[rownames(mat),]$cpg_status) , ]
set.seed(1)
data = data[sort(sample(1:nrow(data), min(nrow(data), 500))), ]
rownames(data) = feats[rownames(data),]$cpg_status
colors=c("black",  "red")
cols = colorRampPalette(colors)(20)
foo = image(t(data), col=cols, xaxt="n", yaxt="n", main="CpG density profiles of 500 random features", useRaster=TRUE)
axis(1, c(0,.5, 1), c(paste0("TSS -", ud_str/1000, "kb"), "TSS", paste0("TSS +", ud_str/1000, "kb")))
axis(2, c(
  0,
  sum(rownames(data)=="poor")/nrow(data)/2, 
  sum(rownames(data)=="poor")/nrow(data), 
  sum(rownames(data)=="poor")/nrow(data) + sum(rownames(data)=="rich")/nrow(data)/2,  
  1), c("", paste0("CpG poor ()", sum(rownames(data)=="poor"), ")"), "", paste0("CpG rich (", sum(rownames(data)=="rich"), ")"), ""), tick=FALSE)
fig_label("D", cex=3)
abline(h=sum(rownames(data)=="poor")/nrow(data), col="white")
# gplots::heatmap.2(data, Rowv=NULL,  Colv=NULL,  dendrogram="none", trace="none", col=cols, main=paste0("CpG enrichment (feats +/-", ud_str/1000, "kb, ", nrow(data), " rnd feats)"), xaxt="n", mar=c(8,5))
```


 
**Clustering of features according to CpG density of genomique region surounding the transcription start site (TSS).**
For each feature we compute the CpG density of the TSS +/-`r ud_str/1000`kb region by counting the number of CG motif in the DNA sequence divided by the length of the sequence. **Panel A** - The distribution of these values (black curve) shows 2 populations, when the distribution of CpG density of 1000 random genomic regions of `r 2*ud_str/1000`kb (blue curve) shows 1 population. The CpG density of random regions orverlaps the feature population with the lowest CpG density. We qualify this population as *CpG poor* features, by opposition to the second population qualify as *CpG rich* feature. In order to separate these two population we i) perform a PCA on the features and ii) apply an EM-clustering algorithm on the first principal component. 
To perfom PCA on features we split genomic region of TSS +/- `r ud_str/1000`kb into `r stp/1000`kb bins. We compute the CpG density of each bin, then we obtain a CpG density matrix of `r nrow(mat)` features and `r ncol(mat)` bins. The heatmap of the panel D shows a 500 random feature sampling of this matrix. **Panel B** - We identify two clusters by using  an esperance maximisation clustering on first principal component of PCA which representing 36,9% of variance. The main graphics shows this clutering with first principal component on x axis and second principal component on y axis, red dots are CpG rich and black dots are CpG poor. The marginals graphics show respectively (same color code of main graphic) on top first principal component distribution clusters (dotted grey vertical line positions treshold between two clusters), on left second principal component distribution for same clusters. **Panel C** - The average CpG density profile of the `r length(idx_rich)` CpG rich features (resp. `r length(idx_poor)` CpG poor features) cluster is plot in red (resp. in black), x axis is the position on the sequence around TSS and y axis is the CpG density. **Panel D** Same information dispatch on heatmap. 
