---
title: "Step 2: Methylation status in healthy tissues"
author: "Florence Pittion, Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---



```{r, echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide", warning=FALSE)
species="Homo_sapiens"
version="hg38"
```

# Step 2 : Methylation status in healthy tissues

CpG-rich associated genes are clustered according to their methylation status in a publicly  available  dataset  of  normal  tissues  (GSE56515,  GSE48472,  GSE64096,  GSE50192,  GSE31848,GSE73375).


## Genes with CpG rich feats

```{r}
if (!exists("feats")) {feats = readRDS(paste0("~/projects/genes/bed_grch38_epimeddb.rds"))}
if (!exists("reducer_func2_name")) {reducer_func2_name = "mean"}
```

```{r echo=TRUE, results="verbatim"}
table(feats$cpg_status)
gs_cpgrich = rownames(feats)[!is.na(feats$cpg_status) & feats$cpg_status=="rich"]
```


```{r}
dim(feats)
gs = rownames(feats)

CpG_dens = function() {
  layout(1, respect=TRUE)
  plot(density(feats$cpg_density, na.rm=TRUE), main="Distribution of CpG density", lwd=3, col="grey", lty=2)
  den_rich = density(feats[feats$cpg_status%in%"rich",]$cpg_density, na.rm=TRUE)
  den_poor = density(feats[feats$cpg_status%in%"poor",]$cpg_density, na.rm=TRUE)
  lines(den_rich$x, den_rich$y * (den_rich$n/nrow(feats)), col=2, lwd=2)
  lines(den_poor$x, den_poor$y * (den_poor$n/nrow(feats)), col=4, lwd=2)
  legend("topright", lty=c(1, 1, 2), col=c("red","blue", "grey"), c("rich", "poor", "both"), lwd=c(2,2,3))
}

CpG_dens()

# pdf(paste0("fig_CpG_dens.pdf"), width=15, height=5)
# CpG_dens()
# dev.off()
```










**Promoter methylation status in normal tissues**

```{r loading_study_meth_all_human_tissues_grch38}
s_meth = mreadRDS("~/projects/all_human_tissues/results/meth/study_meth_all_human_tissues_grch38.rds")
```

```{r process_study_meth_all_human_tissues_grch38, }
feat_indexed_probes = mget_feat_indexed_probes(feats_bed6=feats[gs_cpgrich,], s_meth$platform[,c("Chromosome", "Start")], up_str=up_str, dwn_str=dwn_str)
feat_indexed_probes = feat_indexed_probes[sapply(feat_indexed_probes, length)>5]


tmp_len = sapply(feat_indexed_probes, length)
tmp_len
barplot(table(tmp_len))


#feats$nb_probes =  NA
#feats[names(tmp_len),]$nb_probes = tmp_len
feats$nb_probes = tmp_len[rownames(feats)]
dim(feats)
length(tmp_len)

```


```{r label="reduction of signal by tissue"}
get_meth_by_tissues_by_gene = function (feats, feat_indexed_probes, s_meth, key, reducer_func1, reducer_func2) {
  print("reduction of signal by tissue then by gene")
  print(table(s_meth$exp_grp[[key]]))
  tissues = na.omit(sort(unique(s_meth$exp_grp[[key]])))

  tmp_meth_data = s_meth$data[unique(unlist(feat_indexed_probes)),]
  meth_by_tissues = sapply(tissues, function(tissue) {
    # tissue = tissues[1]
    print(tissue)
    sample_idx = rownames(s_meth$exp_grp)[s_meth$exp_grp[[key]] %in% tissue]
    if (length(sample_idx) > 1) {
      tmp_meth_by_tissues = epimedtools::monitored_apply(mod=50000, tmp_meth_data[,sample_idx], 1, reducer_func1, na.rm=TRUE)      
    } else {
      tmp_meth_by_tissues = reducer_func1(tmp_meth_data[,sample_idx], na.rm=TRUE)
    }
    return(tmp_meth_by_tissues)
  })
  head(meth_by_tissues)
  dim(meth_by_tissues)
  sum(is.na(meth_by_tissues))
  tmp_meth_data = meth_by_tissues
  dim(tmp_meth_data)
  meth_by_tissues_by_gene = epimedtools::monitored_apply(mod=10, feats[names(feat_indexed_probes),], 1, function(gene) {
    # gene = genes[names(gene_indexed_probes),][1,]
    # print(gene[[4]])
    probe_idx = feat_indexed_probes[[gene[[4]]]]
    if (length(probe_idx) > 1) {
      tmp_meth_by_tissues_by_gene = apply(tmp_meth_data[probe_idx,], 2, reducer_func2, na.rm=TRUE)    
    } else {
      tmp_meth_by_tissues_by_gene  = sapply(tmp_meth_data[probe_idx,], reducer_func2, na.rm=TRUE) 
    }
    return(tmp_meth_by_tissues_by_gene)
  })
  meth_by_tissues_by_gene = t(meth_by_tissues_by_gene)
  dim(meth_by_tissues_by_gene)
  sum(is.na(meth_by_tissues_by_gene))    
  return(meth_by_tissues_by_gene)
}
  
if (!exists("mget_meth_by_tissues_by_gene")) {mget_meth_by_tissues_by_gene = memoise::memoise(get_meth_by_tissues_by_gene)}
reducer_func1 = mean
reducer_func2 = get(reducer_func2_name)
key = "tissue_group_level1"
meth_by_tissues_by_gene = mget_meth_by_tissues_by_gene(feats, feat_indexed_probes, s_meth, key, reducer_func1, reducer_func2)

tmp = meth_by_tissues_by_gene
tmp[is.infinite(tmp)] = NA
meanmeth_by_gene = apply(tmp, 1, mean, na.rm=TRUE)

feats$meanmeth = NA
feats[names(meanmeth_by_gene),]$meanmeth = meanmeth_by_gene
head(feats)
```

From Illumina 450k studies, we obtain a methylation matrix of `r ncol(s_meth$data)` samples, 
dispatched into `r length(unique(s_meth$exp_grp$tissue_group_level1))` tissues 
including `r unique(s_meth$exp_grp$tissue_group_level1)`.

This matrix is successively reduced, first by tissues, then by genes with CpG rich promoter. 
The function used to reduce the data is the `mean`.

It means that for each methylation probe, we compute the mean methylation value by tissue.
Then for each genes promoter group of probes, we compute the mean of mean methylation value by tissue.

We obtain an `r nrow(meth_by_tissues_by_gene)` genes by `r ncol(meth_by_tissues_by_gene)` tissues mean of mean methylation value matrix.

Heatmap on the methylation value matrix clearly shows a subset of genes with highly methylated promoter.



## Figure


```{r label="clustering and heatmap", fig.height=9}
data = meth_by_tissues_by_gene
dim(data)
sum(apply(is.na(data), 1, any))
data = data[!apply(is.na(data), 1, any), ]
data = data[!apply(is.infinite(data), 1, any), ]
dim(data)

# clustering base on correlation for tissues
tmp_d = data
tmp_d = t(tmp_d) - apply(tmp_d, 2, mean, na.rm=TRUE)
tmp_d = t(tmp_d)
tmp_d = cor(tmp_d, method="pe")
dim(tmp_d)
hc_col = hclust(dist(1 - tmp_d), method="complete")
Colv = as.dendrogram(hc_col)
dendrogram="col"      

# clustering base on eucl. dist. for genes
d = dist(data)
hc_row = hclust(d, method="complete")
Rowv = as.dendrogram(hc_row)
dendrogram="both"      

# col
colors=c("cyan", "black", "red")
cols = colorRampPalette(colors)(20)

hm = gplots::heatmap.2(data, Rowv=Rowv, Colv=Colv, dendrogram=dendrogram, trace="none", col=cols, main=paste0(reducer_func2_name, " of mean (", nrow(data), " genes x ", ncol(data), " tissues)"), mar=c(10,5), useRaster=TRUE)
idx1 = rownames(data)[unlist(hm$rowDendrogram[[1]])]
idx2 = rownames(data)[unlist(hm$rowDendrogram[[2]])]

feats$methplusplus  = NA
if (mean(data[idx1,]) < mean(data[idx2,])) {
  gs_methplusplus = rownames(data)[unlist(hm$rowDendrogram[[2]])]
  # gs_methminusminus = rownames(data)[unlist(hm$rowDendrogram[[1]])]
} else {
  gs_methplusplus = rownames(data)[unlist(hm$rowDendrogram[[1]])]
  # gs_methminusminus = rownames(data)[unlist(hm$rowDendrogram[[2]])]
}
length(gs_methplusplus)
gs_name = "methplusplus"

feats[[gs_name]] = rownames(feats) %in% gs_methplusplus

```



**CpG-rich genes methylation status in healthy tissues.**
CpG-rich associated genes are clustered according to their methylation status using publicly  available  datasets  of genome-wide DNA methylation data in  normal  tissues obtained whith the Illumina 450k array (GSE56515,  GSE48472,  GSE64096,  GSE50192,  GSE31848, GSE73375). These values are included into a methylation matrix of `r ncol(s_meth$data)` samples, 
dispatched into `r length(unique(s_meth$exp_grp$tissue_group_level1))` tissues. This matrix is successively reduced, first by tissues, then by genes with CpG rich promoter. The function used to reduce the data is the `mean`. A matrix of average methylation values is obtained for `r nrow(data)` genes and `r ncol(meth_by_tissues_by_gene)` tissues. The heatmap shows the results of hierarchical clustering on this matrix with: high methylated values in red and low methylation values in blue. The tissues are presented on the x axis (with a hierarchical clustering dendrogram on tissues on the top of the panel) and the genes are presented on the y axis (with the hierarchical clustering dendrogram of the genes on the left of the panel). The first branch of this genes dendrogram selects a subset of `r sum(feats[[gs_name]])` CpG-rich genes highly and widely methylated in most healthy tissues.




```{r selecting_genes, echo=TRUE, results="verbatim", eval=TRUE}
table(feats$cpg_status, feats[[gs_name]], useNA="ifany")
# expression_in_gtex(rownames(feats)[feats[["methplusplus"]]])
```












