---
title: "Step 4: Results"
subtitle: "Hunting DNA methylation repression of transcriptional activity"
author: "Florence Pittion, Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---

# Step 4 : Results

```{r, echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide", warning=FALSE)
```

## Figure

```{r}
if (! exists("momik_pvals")) {
  genes_singleton = mget_genes(tcga_project)
  genes = genes_singleton$genes

  s_meth = mreadRDS("~/projects/all_human_tissues/results/meth/study_meth_all_human_tissues_grch38.rds")
  feat_indexed_probes = mget_feat_indexed_probes(feats_bed6=genes, s_meth$platform[,c("Chromosome", "Start")], up_str=ud_str, dwn_str=2500)
  tmp_gs = intersect(names(feat_indexed_probes), rownames(genes))
  momik_pvals = epimedtools::monitored_apply(mod=10, t(t(tmp_gs)), 1, function(gene_symbol) {
    #print(gene_symbol)
    feat = genes[gene_symbol,]
    tss = ifelse(feat[[6]]=="+", feat[[2]], feat[[3]])
    interaction_range=2500
    region_id = paste0(feat[[1]], ":", tss-interaction_range, "-", tss+interaction_range)      
    feat_indexed_probes[[region_id]] = feat_indexed_probes[[gene_symbol]]
    mdata = try(mget_multiomic_data(region_id=region_id, gene_symbols=gene_symbol, tcga_project=tcga_project, feat_indexed_probes=feat_indexed_probes))   
    if (class(mdata) != "try-error"){
      if ((!is.null(mdata$d)) & (gene_symbol %in% colnames(mdata$d))) {
        mdata$d 
        head(mdata$d)
        Y=mdata$d[,gene_symbol]
        X=apply(mdata$d[,mdata$probes],1,mean, na.rm=TRUE)
        m= lm(X~Y)
        sm=summary(m)
        fstat=sm$fstatistic
        pval = pf(fstat[[1]], fstat[[2]], fstat[[3]], lower.tail = FALSE)
        return(pval)
      } else {
        warning(paste0("probleme with ", gene_symbol))
        return(NULL)
      }
    } else {
      warning(paste0("probleme with ", gene_symbol))
      return(NULL)
    }
  })  
  names(momik_pvals) = tmp_gs  
  momik_pvals = unlist(momik_pvals)
}
feats$momik_pvals = NA
feats[names(momik_pvals),]$momik_pvals = momik_pvals


if (! exists("survival_pvalhrs")) {
  s = mreadRDS(paste0("tcga_studies/study_", tcga_project, "_trscr.rds"))
  s = truncate_survival(s, 60)  

  genes_singleton = mget_genes(tcga_project)
  genes = genes_singleton$genes

  tmp_gs = intersect(rownames(feats), rownames(s$data))
  survival_pvalhrs = epimedtools::monitored_apply(mod=10, t(t(tmp_gs)), 1, function(gene_symbol) {
    idx_sample = rownames(s$exp_grp)[!is.na(s$exp_grp$os)]
    ss = s$exp_grp[idx_sample,]$os
    v = s$data[gene_symbol, idx_sample]
    pv = epimedtools::coxres(ss=ss, v=v)[[1]]
    hr = epimedtools::coxres(ss=ss, v=v)[[4]]
    ret = c(pv=pv, hr=hr)
    return(ret)
  })  
  survival_pvalhrs = t(survival_pvalhrs)
  rownames(survival_pvalhrs) = tmp_gs
  survival_pvalhrs[is.na(survival_pvalhrs[,2]),2] = 1
  survival_hrs = unlist(survival_pvalhrs[,2])
  survival_pvals = unlist(survival_pvalhrs[,1])
}
feats$survival_pvals = NA
feats[rownames(survival_pvalhrs),]$survival_pvals = survival_pvalhrs[,1]
feats$survival_hrs = NA
feats[rownames(survival_pvalhrs),]$survival_hrs = survival_pvalhrs[,2]

# layout(matrix(1:2,1), respect=TRUE)
# plot(feats$survival_hrs, -log10(feats$survival_pvals))
# idx = rownames(feats)[feats[[gs_name]]]
# points(feats[idx,]$survival_hrs, -log10(feats[idx,]$survival_pvals), col="red")



# plot(-log10(feats$survival_pvals), -log10(feats$momik_pvals), pch=ifelse(feats$cpg_status %in% "rich", "black", "grey"), col=adjustcolor(1, alpha.f=.3), pch=".")

ntop=100
idx = rownames(feats)[feats[[gs_name]]]
idx2 = idx[order(feats[idx,]$l2fc_GSE45332, decreasing=TRUE)][1:ntop]
idx3 = idx[order(feats[idx,]$l2fc_GSE5816, decreasing=TRUE)][1:ntop]
gene_symbols = intersect(intersect(idx[feats[idx,]$survival_pvals < 0.05], idx[-log10(feats[idx,]$momik_pvals)>5]), union(idx2, idx3)) 
gene_symbols = gene_symbols[order(feats[gene_symbols,]$momik_pvals)][1:min(20, length(gene_symbols))]


layout(1, respect=TRUE)
plot(-log10(feats$momik_pvals), -log10(feats$survival_pvals), col=adjustcolor(1, alpha.f=.3), xlab="multi-omic score", ylab="-log10(pval_logrank)", xlim=c(0,1.5*max(-log10(feats[gene_symbols,]$momik_pvals))), ylim=c(0,1.5*max(-log10(feats[gene_symbols,]$survival_pvals))))
abline(h=-log10(0.05), lty=2)
points(-log10(feats[idx,]$momik_pvals), -log10(feats[idx,]$survival_pvals), col="red")
points(-log10(feats[idx2,]$momik_pvals), -log10(feats[idx2,]$survival_pvals), col=adjustcolor("blue", alpha.f=.8), pch=16)
points(-log10(feats[idx3,]$momik_pvals), -log10(feats[idx3,]$survival_pvals), col=adjustcolor("green", alpha.f=.8), pch=16)

text(-log10(feats[gene_symbols,]$momik_pvals), -log10(feats[gene_symbols,]$survival_pvals), gene_symbols, pos=3, col="white")
text(-log10(feats[gene_symbols,]$momik_pvals), -log10(feats[gene_symbols,]$survival_pvals), gene_symbols, pos=3, col=adjustcolor(1, alpha.f=0.8))
```





```{r survival, eval=TRUE}
tcga_project = "TCGA-LUSC"
s = mreadRDS(paste0("tcga_studies/study_", tcga_project, "_trscr.rds"))
s = truncate_survival(s, 60)


for (gene_symbol in gene_symbols) {
  idx_sample = rownames(s$exp_grp)[!is.na(s$exp_grp$os)]
  ss = s$exp_grp[idx_sample,]$os
  v = s$data[gene_symbol, idx_sample]
  try(epimedtools::plot_survival_panel_simple2(ss=ss, v=v, main=gene_symbol))
  try(momic_pattern(gene_symbol, tcga_project))
}
```






## Export features


```{r}
feat_export_filename = paste0("feats_", prefix3, ".xlsx")
```

Annotated features are exported as [`r feat_export_filename`](`r feat_export_filename`) file. 

```{r results="verbatim"}
head(feats)
```

```{r export_genes}
WriteXLS::WriteXLS(feats, feat_export_filename)
```



```{r core feats, eval=FALSE}
## Core feats
# plot 2 firsts studies
#threshold
#study 1
i = 1
#study 2
j = 1
layout(1, respect=TRUE)
plot(feats$l2fc_GSE45332, feats$l2fc_GSE5816, xlab="l2fc GSE45332 WTvsDKO", ylab="l2fc GSE5816", pch=".", col="grey")
idx = rownames(feats)[!is.na(feats$methplusplus) & feats$methplusplus]
points(feats[idx,]$l2fc_GSE45332, feats[idx,]$l2fc_GSE5816, col=2, pch=1)
abline(h=i, col=1, lty=2)
abline(v=j, col=1, lty=2)
legend("topleft", pch=1, col=2, "gs_methplusplus")

# features of interest
corefeats = rownames(feats)[
  !is.na(feats$l2fc_GSE45332) & feats$l2fc_GSE45332>i & 
  !is.na(feats$l2fc_GSE5816)  & feats$l2fc_GSE5816>j  & 
  !is.na(feats$methplusplus)  & feats$methplusplus
]
length(corefeats) 
```


```{r results="verbatim", eval=FALSE}
# Annotated core features are exported as [`r corefeat_export_filename`](`r corefeat_export_filename`) file. 
if (length(corefeats!=0)){
  corefeat_export_filename = paste0("corefeats_", prefix3, ".xlsx")
  head(corefeats)
  tmp_d = feats[corefeats,]
  WriteXLS::WriteXLS(tmp_d, corefeat_export_filename)
}
```




```{r gtex, fig.height=9, eval=FALSE}
## Expression in GTEx
expression_in_gtex(gene_symbols)
```






