---
title: "Step 4: Results"
subtitle: "Hunting DNA methylation repression of transcriptional activity"
author: "Florence Pittion, Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---

# Step 4 : Results

```{r, echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide", warning=FALSE)
```

## Multiomic

```{r}
if (! exists("momik_pvalbetas")) {
  genes_singleton = mget_genes(tcga_project)
  genes = genes_singleton$genes

  s_meth = mreadRDS("~/projects/all_human_tissues/results/meth/study_meth_all_human_tissues_grch38.rds")
  feat_indexed_probes = mget_feat_indexed_probes(feats_bed6=genes, s_meth$platform[,c("Chromosome", "Start")], up_str=ud_str, dwn_str=2500)
  tmp_gs = intersect(names(feat_indexed_probes), rownames(genes))
  momik_pvalbetas = epimedtools::monitored_apply(mod=10, t(t(tmp_gs)), 1, function(gene_symbol) {
    #print(gene_symbol)
    feat = genes[gene_symbol,]
    tss = ifelse(feat[[6]]=="+", feat[[2]], feat[[3]])
    interaction_range=2500
    region_id = paste0(feat[[1]], ":", tss-interaction_range, "-", tss+interaction_range)      
    feat_indexed_probes[[region_id]] = feat_indexed_probes[[gene_symbol]]
    mdata = try(mget_multiomic_data(region_id=region_id, gene_symbols=gene_symbol, tcga_project=tcga_project, feat_indexed_probes=feat_indexed_probes))   
    if (class(mdata) != "try-error"){
      if ((!is.null(mdata$d)) & (gene_symbol %in% colnames(mdata$d))) {
        mdata$d 
        head(mdata$d)
        Y=mdata$d[,gene_symbol]
        X=apply(mdata$d[,mdata$probes],1,mean, na.rm=TRUE)
        m = lm(X~Y)
        
        sm =summary(m)
        fstat=sm$fstatistic
        pval = pf(fstat[[1]], fstat[[2]], fstat[[3]], lower.tail = FALSE)
        beta = m$coefficient[[2]]
        return(c(beta=beta, pval=pval))
      } else {
        warning(paste0("probleme with ", gene_symbol))
        return(NULL)
      }
    } else {
      warning(paste0("probleme with ", gene_symbol))
      return(NULL)
    }
  })  
  names(momik_pvalbetas) = tmp_gs  
  momik_pvalbetas = do.call(rbind,momik_pvalbetas)
}
feats$momik_pvals = NA
feats[rownames(momik_pvalbetas),]$momik_pvals = momik_pvalbetas[,2]
feats$momik_betas = NA
feats[rownames(momik_pvalbetas),]$momik_betas = momik_pvalbetas[,1]
feats$momik_lpvals = -log10(feats$momik_pvals)


if (! exists("survival_pvalhrs")) {
  s = mreadRDS(paste0("tcga_studies/study_", tcga_project, "_trscr.rds"))
  s = truncate_survival(s, 60)  

  genes_singleton = mget_genes(tcga_project)
  genes = genes_singleton$genes

  tmp_gs = intersect(rownames(feats), rownames(s$data))
  survival_pvalhrs = epimedtools::monitored_apply(mod=10, t(t(tmp_gs)), 1, function(gene_symbol) {
    # gene_symbol = "MIR6723"
    # print(gene_symbol)
    idx_sample = rownames(s$exp_grp)[!is.na(s$exp_grp$os)]
    ss = s$exp_grp[idx_sample,]$os
    v = s$data[gene_symbol, idx_sample]
    pv = try(epimedtools::coxres(ss=ss, v=v)[[1]])
    if (class(pv)!="try-error") {
      hr = epimedtools::coxres(ss=ss, v=v)[[4]]
    } else {
      pv = 1
      hr = 1
    }
    ret = c(hr=hr, pv=pv)
    return(ret)
  })  
  survival_pvalhrs = t(survival_pvalhrs)
  rownames(survival_pvalhrs) = tmp_gs
  survival_pvalhrs[is.na(survival_pvalhrs[,1]),1] = 1
}
feats$survival_pvals = NA
feats[rownames(survival_pvalhrs),]$survival_pvals = survival_pvalhrs[,2]
feats$survival_hrs = NA
feats[rownames(survival_pvalhrs),]$survival_hrs = survival_pvalhrs[,1]
feats$survival_lpvals = -log10(feats$survival_pvals)

ntop=100
idx = rownames(feats)[feats[[gs_name]]]
idx2 = idx[order(feats[idx,]$l2fc_GSE45332, decreasing=TRUE)][1:ntop]
idx3 = idx[order(feats[idx,]$l2fc_GSE5816, decreasing=TRUE)][1:ntop]


```

## Figure

```{r fig.width=15, fig.height=15}
gene_symbols = (union(idx2, idx3)) 
gene_symbols = gene_symbols[order(feats[gene_symbols,]$momik_pvals)][1:min(6, length(gene_symbols))]

layout_mat2 = cbind(matrix(1, nrow=6, ncol=5), c(2:7))
layout_mat2 
layout(layout_mat2, respect=TRUE)
plot(feats$momik_betas, feats$momik_lpvals, col="grey")
points(feats[idx,]$momik_betas, feats[idx,]$momik_lpvals  , col="red")
points(feats[idx2,]$momik_betas, feats[idx2,]$momik_lpvals, col="blue", pch=16)
points(feats[idx3,]$momik_betas, feats[idx3,]$momik_lpvals, col="purple", pch=16)
text(feats[gene_symbols,]$momik_betas, feats[gene_symbols,]$momik_lpvals, gene_symbols, col=adjustcolor(1, alpha.f=1), pos=c(1,3))

legend("topright", 
  c("genes", paste0(gs_name, " genes"), paste0(gs_name, " genes upregulated GSE45332 DA"), paste0(gs_name, " genes upregulated GSE5816 DA")), 
  pch=c(1, 1, 16, 16), 
  col=c("grey", "red", "blue", "purple"),
  lty=c(0,0,0,0)
)

fig_label("A", cex=3)

gene_set = idx
gs_name = "momik_betas"
expression_vector = feats[,gs_name]
names(expression_vector) = rownames(feats)
expression_vector = expression_vector[!is.na(expression_vector)]
et_gsea_plot(expression_vector, gene_set, prefix=gs_name, nperm=3)
fig_label("B", cex=3)

gs_name = "momik_lpvals"
expression_vector = feats[,gs_name]
names(expression_vector) = rownames(feats)
expression_vector = expression_vector[!is.na(expression_vector)]
et_gsea_plot(expression_vector, gene_set, prefix=gs_name, nperm=3)
fig_label("C", cex=3)

gene_set = idx2
gs_name = "momik_betas"
expression_vector = feats[,gs_name]
names(expression_vector) = rownames(feats)
expression_vector = expression_vector[!is.na(expression_vector)]
et_gsea_plot(expression_vector, gene_set, prefix=gs_name, nperm=3)
fig_label("D", cex=3)

gs_name = "momik_lpvals"
expression_vector = feats[,gs_name]
names(expression_vector) = rownames(feats)
expression_vector = expression_vector[!is.na(expression_vector)]
et_gsea_plot(expression_vector, gene_set, prefix=gs_name, nperm=3)
fig_label("E", cex=3)

gene_set = idx3
gs_name = "momik_betas"
expression_vector = feats[,gs_name]
names(expression_vector) = rownames(feats)
expression_vector = expression_vector[!is.na(expression_vector)]
et_gsea_plot(expression_vector, gene_set, prefix=gs_name, nperm=3)
fig_label("F", cex=3)

gs_name = "momik_lpvals"
expression_vector = feats[,gs_name]
names(expression_vector) = rownames(feats)
expression_vector = expression_vector[!is.na(expression_vector)]
et_gsea_plot(expression_vector, gene_set, prefix=gs_name, nperm=3)
fig_label("G", cex=3)
```


```{r}
tcga_project = "TCGA-LUSC"
s = mreadRDS(paste0("tcga_studies/study_", tcga_project, "_trscr.rds"))
s = truncate_survival(s, 60)
idx_sample = rownames(s$exp_grp)[!is.na(s$exp_grp$os)]


foo = sapply(gene_symbols, function(gene_symbol) {
  ss = s$exp_grp[idx_sample,]$os
  v = s$data[gene_symbol, idx_sample]
  par(mar=c(2.5, 2.5, 4.1, 2.1))
  try(momic_pattern(gene_symbol, tcga_project))
  ret
})
```


## Survival

```{r survival, eval=TRUE}
gene_symbols = intersect(intersect(idx[feats[idx,]$survival_pvals < 0.05], idx[-log10(feats[idx,]$momik_pvals)>5]), union(idx2, idx3)) 
gene_symbols = gene_symbols[order(feats[gene_symbols,]$momik_pvals)][1:min(20, length(gene_symbols))]


layout(1, respect=TRUE)
plot(-log10(feats$momik_pvals), -log10(feats$survival_pvals), col=adjustcolor(1, alpha.f=.3), 
  xlab="multiomic score", 
  ylab="-log10(survival logrank p-value)", xlim=c(0,2*max(-log10(feats[gene_symbols,]$momik_pvals))), ylim=c(0,2*max(-log10(feats[gene_symbols,]$survival_pvals))))
abline(h=-log10(0.05), lty=2)
points(-log10(feats[idx,]$momik_pvals), -log10(feats[idx,]$survival_pvals), col="red")
points(-log10(feats[idx2,]$momik_pvals), -log10(feats[idx2,]$survival_pvals), col=adjustcolor("blue", alpha.f=.8), pch=16)
points(-log10(feats[idx3,]$momik_pvals), -log10(feats[idx3,]$survival_pvals), col=adjustcolor("purple", alpha.f=.8), pch=16)

# text(-log10(feats[gene_symbols,]$momik_pvals), -log10(feats[gene_symbols,]$survival_pvals), gene_symbols, pos=3, col="white")
text(-log10(feats[gene_symbols,]$momik_pvals), -log10(feats[gene_symbols,]$survival_pvals), gene_symbols, col=adjustcolor(1, alpha.f=1), pos=c(1,3))
legend("topright", 
  c("genes", paste0(gs_name, " genes"), paste0(gs_name, " genes upregulated GSE45332 DA"), paste0(gs_name, " genes upregulated GSE5816 DA"), "5% survival logrank p-value threshold "), 
  pch=c(1, 1, 16, 16, NA), 
  col=c("grey", "red", "blue", "purple", "black"),
  lty=c(0,0,0,0,2)
)


tcga_project = "TCGA-LUSC"
s = mreadRDS(paste0("tcga_studies/study_", tcga_project, "_trscr.rds"))
s = truncate_survival(s, 60)
idx_sample = rownames(s$exp_grp)[!is.na(s$exp_grp$os)]


foo = sapply(gene_symbols, function(gene_symbol) {
  ss = s$exp_grp[idx_sample,]$os
  v = s$data[gene_symbol, idx_sample]
  par(mar=c(2.5, 2.5, 4.1, 2.1))
  ret = try(epimedtools::plot_survival_panel_simple2(ss=ss, v=v, main=gene_symbol))
  try(momic_pattern(gene_symbol, tcga_project))
  ret
})

# thresh = do.call(rbind,foo[1,])[,2]
#
# idx_genes = gene_symbols[1:4]
# v = apply(s$data[idx_genes,idx_sample] > thresh[idx_genes], 2, sum)
# table(v)
# ss = s$exp_grp[idx_sample,]$os
# par(mar=c(2.5, 2.5, 4.1, 2.1))
# ret = try(epimedtools::plot_survival_panel_simple2(ss=ss, v=v, main=gene_symbol))

```

## Discussion

Nicotine promotes the development of non-small cell lung cancer through activating LINC00460 and PI3K/Akt signaling.
Zhao H, Wang Y, Ren X.
Biosci Rep. 2019 Jun 7;39(6):BSR20182443. doi: 10.1042/BSR20182443. Print 2019 Jun 28.
PMID: 31123168 Free PMC article.

LINC00460 promotes proliferation and inhibits apoptosis of non-small cell lung cancer cells through targeted regulation of miR-539.
Wang HX, Kang LJ, Qin X, Xu J, Fei JW.
Eur Rev Med Pharmacol Sci. 2020 Jun;24(12):6752-6758. doi: 10.26355/eurrev_202006_21663.
PMID: 32633366 Free article.

Long non-coding RNA linc00460 promotes epithelial-mesenchymal transition and cell migration in lung cancer cells.
Li K, Sun D, Gou Q, Ke X, Gong Y, Zuo Y, Zhou JK, Guo C, Xia Z, Liu L, Li Q, Dai L, Peng Y.
Cancer Lett. 2018 Apr 28;420:80-90. doi: 10.1016/j.canlet.2018.01.060. Epub 2018 Jan 31.
PMID: 29409808






## Export features


```{r}
feat_export_filename = paste0("feats_", prefix3, ".xlsx")
```

Annotated features are exported as [`r feat_export_filename`](`r feat_export_filename`) file. 

```{r results="verbatim"}
head(feats)
```

```{r export_genes}
WriteXLS::WriteXLS(feats, feat_export_filename)
```



```{r core feats, eval=FALSE}
## Core feats
# plot 2 firsts studies
#threshold
#study 1
i = 1
#study 2
j = 1
layout(1, respect=TRUE)
plot(feats$l2fc_GSE45332, feats$l2fc_GSE5816, xlab="l2fc GSE45332 WTvsDKO", ylab="l2fc GSE5816", pch=".", col="grey")
idx = rownames(feats)[!is.na(feats$methplusplus) & feats$methplusplus]
points(feats[idx,]$l2fc_GSE45332, feats[idx,]$l2fc_GSE5816, col=2, pch=1)
abline(h=i, col=1, lty=2)
abline(v=j, col=1, lty=2)
legend("topleft", pch=1, col=2, "gs_methplusplus")

# features of interest
corefeats = rownames(feats)[
  !is.na(feats$l2fc_GSE45332) & feats$l2fc_GSE45332>i & 
  !is.na(feats$l2fc_GSE5816)  & feats$l2fc_GSE5816>j  & 
  !is.na(feats$methplusplus)  & feats$methplusplus
]
length(corefeats) 
```


```{r results="verbatim", eval=FALSE}
# Annotated core features are exported as [`r corefeat_export_filename`](`r corefeat_export_filename`) file. 
if (length(corefeats!=0)){
  corefeat_export_filename = paste0("corefeats_", prefix3, ".xlsx")
  head(corefeats)
  tmp_d = feats[corefeats,]
  WriteXLS::WriteXLS(tmp_d, corefeat_export_filename)
}
```




```{r gtex, fig.height=9, eval=FALSE}
## Expression in GTEx
expression_in_gtex(gene_symbols)
```






